In [182]:
import pandas as pd
pd.set_option('display.max_columns', 1500)

import numpy as np

import warnings
warnings.filterwarnings('ignore')

#Extend cell width
from IPython.core.display import display, HTML
display(HTML("<style>.container { width:80% !important; }</style>"))

## Load files into Pandas DF

In [183]:
student_df = pd.read_csv(filepath_or_buffer="/Users/aishwaryaadiki/Downloads/student-mat.csv", sep=",", header='infer')
nutr_df = pd.read_csv(filepath_or_buffer="/Users/aishwaryaadiki/Downloads/Nutritional Data for Fast Foods.csv", sep=",", header='infer')

In [184]:
student_df.head()

school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher   
1     GP   F   17       U     GT3       T     1     1  at_home     other   
2     GP   F   15       U     LE3       T     1     1  at_home     other   
3     GP   F   15       U     GT3       T     4     2   health  services   
4     GP   F   16       U     GT3       T     3     3    other     other   

   reason guardian  traveltime  studytime  failures schoolsup famsup paid  \
0  course   mother           2          2         0       yes     no   no   
1  course   father           1          2         0        no    yes   no   
2   other   mother           1          2         3       yes     no  yes   
3    home   mother           1          3         0        no    yes  yes   
4    home   father           1          2         0        no    yes  yes   

  activities nursery higher internet romantic  famrel  freetime  goout  Dalc  \
0         no     yes    yes       no       no       4         3      4     1   
1         no      no    yes      yes       no       5         3      3     1   
2         no     yes    yes      yes       no       4         3      2     2   
3        yes     yes    yes      yes      yes       3         2      2     1   
4         no     yes    yes       no       no       4         3      2     1   

   Walc  health  absences grade  
0     1       3         6     F  
1     1       3         4     F  
2     3       3        10     C  
3     1       5         2     B  
4     2       5         4     C

In [185]:
student_df.shape

(395, 31)

In [186]:
nutr_df.head()

Unnamed: 0 Fast Food Restaurant                         Item    Type  \
0           0           McDonald's                    Hamburger  Burger   
1           1           McDonald's                 Cheeseburger  Burger   
2           2           McDonald's                            ?  Burger   
3           3           McDonald's  Quarter Pounder with Cheese  Burger   
4           4           McDonald's       Bacon Clubhouse Burger  Burger   

   Serving Size (g)  Calories  Total Fat (g)  Saturated Fat (g)  \
0                98       240            8.0                3.0   
1               113       290           11.0                5.0   
2               211       530           27.0               10.0   
3               202       520           26.0               12.0   
4               270       720           40.0               15.0   

   Trans Fat (g)  Sodium (mg)  Carbs (g)  Sugars (g)  Protein (g)  
0            0.0        480.0       32.0         6.0         12.0  
1            0.5        680.0       33.0         7.0         15.0  
2            1.0        960.0       47.0         9.0         24.0  
3            1.5       1100.0       41.0        10.0         30.0  
4            1.5       1470.0       51.0        14.0         39.0

### (Part a) average age of the students in this dataset
### (Part b) average freetime in this dataset

In [187]:
student_df.loc[:, 'age'].mean()

16.696202531645568

In [188]:
student_df.loc[:, 'freetime'].mean()

3.2354430379746835

The average age of the students in this dataset is 16.70 years(4 sig. figs) 
The average freetime in this dataset is 3.235 hours (4 sig. figs)

In [189]:
gradeA = len(student_df[student_df['grade'] == 'A'])
gradeF = len(student_df[student_df['grade'] == 'F'])

gradeA/gradeF

0.7681159420289855


### The ratio of Grade 'A' to Grade 'F' in this dataset is 0.7681

In [195]:
student_df[(student_df['grade'] <= 'B')].groupby('sex').describe()



age                                                      Medu            \
    count       mean       std   min   25%   50%    75%   max count      mean   
sex                                                                             
F    74.0  16.797297  1.170174  15.0  16.0  17.0  17.75  20.0  74.0  2.945946   
M    82.0  16.365854  1.262099  15.0  15.0  16.0  17.00  20.0  82.0  3.048780   

                                        Fedu                                \
          std  min  25%  50%  75%  max count      mean       std  min  25%   
sex                                                                          
F    1.084030  0.0  2.0  3.0  4.0  4.0  74.0  2.621622  1.056003  1.0  2.0   
M    1.041158  0.0  2.0  3.0  4.0  4.0  82.0  2.804878  1.104860  0.0  2.0   

                   traveltime                                               \
     50%  75%  max      count      mean       std  min  25%  50%  75%  max   
sex                                                                          
F    3.0  4.0  4.0       74.0  1.364865  0.631899  1.0  1.0  1.0  2.0  3.0   
M    3.0  4.0  4.0       82.0  1.353659  0.616080  1.0  1.0  1.0  2.0  4.0   

    studytime                                              failures            \
        count      mean       std  min  25%  50%  75%  max    count      mean   
sex                                                                             
F        74.0  2.364865  0.837058  1.0  2.0  2.0  3.0  4.0     74.0  0.148649   
M        82.0  1.902439  0.883452  1.0  1.0  2.0  2.0  4.0     82.0  0.085366   

                                       famrel                                \
          std  min  25%  50%  75%  max  count      mean       std  min  25%   
sex                                                                           
F    0.427879  0.0  0.0  0.0  0.0  2.0   74.0  3.851351  0.916850  1.0  3.0   
M    0.281145  0.0  0.0  0.0  0.0  1.0   82.0  3.987805  1.012196  1.0  4.0   

                   freetime                                               \
     50%  75%  max    count      mean       std  min  25%  50%  75%  max   
sex                                                                        
F    4.0  4.0  5.0     74.0  2.986486  1.026942  1.0  2.0  3.0  4.0  5.0   
M    4.0  5.0  5.0     82.0  3.500000  1.021376  2.0  3.0  3.5  4.0  5.0   

    goout                                               Dalc            \
    count      mean       std  min  25%  50%  75%  max count      mean   
sex                                                                      
F    74.0  2.891892  0.987144  1.0  2.0  3.0  3.0  5.0  74.0  1.202703   
M    82.0  3.012195  1.071446  1.0  2.0  3.0  4.0  5.0  82.0  1.573171   

                                        Walc                                \
          std  min  25%  50%  75%  max count      mean       std  min  25%   
sex                                                                          
F    0.596330  1.0  1.0  1.0  1.0  4.0  74.0  1.878378  0.992474  1.0  1.0   
M    1.042676  1.0  1.0  1.0  2.0  5.0  82.0  2.280488  1.372340  1.0  1.0   

                   health                                               \
     50%  75%  max  count      mean       std  min  25%  50%  75%  max   
sex                                                                      
F    2.0  3.0  5.0   74.0  3.081081  1.541719  1.0  2.0  3.0  5.0  5.0   
M    2.0  3.0  5.0   82.0  3.792683  1.403152  1.0  3.0  4.0  5.0  5.0   

    absences                                                 
       count      mean       std  min   25%  50%  75%   max  
sex                                                          
F       74.0  6.310811  7.494154  0.0  0.25  4.0  9.5  40.0  
M       82.0  3.780488  4.139614  0.0  0.00  2.0  6.0  20.0

### **156 students received a Grade B or higher**

### Consider the grade assignment to GPA below. 
- A: 4
- B: 3
- C: 2
- F: 1

In [196]:
student_df['grade'].replace('A', 4, inplace=True)
student_df['grade'].replace('B', 3, inplace=True)
student_df['grade'].replace('C', 2, inplace=True)
student_df['grade'].replace('F', 1, inplace=True)

(student_df[student_df['sex'] == 'F']).loc[:, 'grade'].mean()



2.2740384615384617

In [197]:
(student_df[student_df['sex'] == 'M']).loc[:, 'grade'].mean()

2.443850267379679

#### **The average GPA for female students is 2.274**
#### **The avergae GPA for male students is 2.444**

In [198]:
nutr_df.columns[nutr_df.isnull().any()]

Index(['Total Fat (g)', 'Trans Fat (g)', 'Sodium (mg)', 'Protein (g)'], dtype='object')

In [199]:
nutr_df.fillna(np.nan, inplace=True)
nutr_df.replace('?', np.nan, inplace=True)


In [200]:
nutr_df

Unnamed: 0 Fast Food Restaurant                         Item  \
0             0           McDonald's                    Hamburger   
1             1           McDonald's                 Cheeseburger   
2             2           McDonald's                          NaN   
3             3           McDonald's  Quarter Pounder with Cheese   
4             4           McDonald's       Bacon Clubhouse Burger   
..          ...                  ...                          ...   
121         121      In-N-Out Burger            Hamburger w/Onion   
122         122      In-N-Out Burger         Cheeseburger w/Onion   
123         123      In-N-Out Burger        Double-Double w/Onion   
124         124      In-N-Out Burger                 French Fries   
125         125      In-N-Out Burger       Chocolate Shake (15oz)   

             Type  Serving Size (g)  Calories  Total Fat (g)  \
0          Burger                98       240            8.0   
1          Burger               113       290           11.0   
2          Burger               211       530           27.0   
3          Burger               202       520           26.0   
4          Burger               270       720           40.0   
..            ...               ...       ...            ...   
121        Burger               243       390           19.0   
122        Burger               268       480           27.0   
123        Burger               330       670           41.0   
124  French Fries               125       395           18.0   
125     Milkshake               321       590           29.0   

     Saturated Fat (g)  Trans Fat (g)  Sodium (mg)  Carbs (g)  Sugars (g)  \
0                  3.0            0.0        480.0       32.0         6.0   
1                  5.0            0.5        680.0       33.0         7.0   
2                 10.0            1.0        960.0       47.0         9.0   
3                 12.0            1.5       1100.0       41.0        10.0   
4                 15.0            1.5       1470.0       51.0        14.0   
..                 ...            ...          ...        ...         ...   
121                5.0            0.0        650.0       39.0        10.0   
122               10.0            0.5       1000.0       39.0        10.0   
123               18.0            1.0       1440.0       39.0        10.0   
124                5.0            0.0        245.0       54.0         0.0   
125               19.0            1.0        320.0       72.0        65.0   

     Protein (g)  
0           12.0  
1           15.0  
2           24.0  
3           30.0  
4           39.0  
..           ...  
121         16.0  
122         22.0  
123         37.0  
124          7.0  
125         10.0  

[126 rows x 13 columns]

#### **How many unique values are in the column `Item`?**

In [201]:
len(nutr_df['Item'].unique())

101

#### **List unique values and counts for the column `Type`**

In [202]:
subarray = nutr_df['Type']
pd.value_counts(subarray)

Type
Burger                      69
Milkshake                   12
French Fries                12
Breaded Chicken Sandwich    11
Grilled Chicken Sandwich    11
Chicken Nuggets             11
Name: count, dtype: int64

#### **Using a groupby statement to find the average calorie for each food type.**


In [203]:
nutr_df[['Type','Calories']].groupby(['Type']).mean()

Calories
Type                                
Breaded Chicken Sandwich  522.090909
Burger                    620.202899
Chicken Nuggets           274.545455
French Fries              314.083333
Grilled Chicken Sandwich  408.000000
Milkshake                 606.666667

#### **The average calories for French Fries is 314.08 cal**


### **Restaurants in the order of highest protein per serving to lowest**

In [205]:
sortedarray = nutr_df.sort_values(by=['Protein (g)'], ascending=False)
subarray = sortedarray[['Protein (g)','Fast Food Restaurant', 'Type']]
answer = subarray[subarray['Type'] == 'Burger'].groupby('Fast Food Restaurant')
answer.head()

Protein (g) Fast Food Restaurant    Type
26          69.0              Wendy's  Burger
55          67.0                Sonic  Burger
115         60.0          Whataburger  Burger
27          57.0              Wendy's  Burger
42          57.0      Jack in the Box  Burger
16          57.0          Burger King  Burger
93          56.0             Hardee's  Burger
114         55.0          Whataburger  Burger
112         49.0          Whataburger  Burger
5           48.0           McDonald's  Burger
25          48.0              Wendy's  Burger
77          48.0           Carl's Jr.  Burger
80          47.0           Carl's Jr.  Burger
113         47.0          Whataburger  Burger
90          45.0             Hardee's  Burger
66          45.0          Dairy Queen  Burger
15          43.0          Burger King  Burger
45          41.0      Jack in the Box  Burger
92          40.0             Hardee's  Burger
81          39.0           Carl's Jr.  Burger
56          39.0                Sonic  Burger
4           39.0           McDonald's  Burger
28          38.0              Wendy's  Burger
67          37.0          Dairy Queen  Burger
123         37.0      In-N-Out Burger  Burger
78          36.0           Carl's Jr.  Burger
46          35.0      Jack in the Box  Burger
57          35.0                Sonic  Burger
91          34.0             Hardee's  Burger
68          34.0          Dairy Queen  Burger
89          32.0             Hardee's  Burger
54          31.0                Sonic  Burger
3           30.0           McDonald's  Burger
24          30.0              Wendy's  Burger
76          29.7           Carl's Jr.  Burger
111         29.0          Whataburger  Burger
43          29.0      Jack in the Box  Burger
65          27.0          Dairy Queen  Burger
13          26.0          Burger King  Burger
64          25.0          Dairy Queen  Burger
2           24.0           McDonald's  Burger
122         22.0      In-N-Out Burger  Burger
103         21.0         White Castle  Burger
53          19.0                Sonic  Burger
121         16.0      In-N-Out Burger  Burger
40          16.0      Jack in the Box  Burger
1           15.0           McDonald's  Burger
102         15.0         White Castle  Burger
12          13.0          Burger King  Burger
101         11.0         White Castle  Burger
11          11.0          Burger King  Burger
100          8.0         White Castle  Burger
99           7.0         White Castle  Burger

#### **The fast-food restaurant that has the Burger with the highest protein per serving is Wendy's coming in at 69.0 grams**

## **H2O-3 cluster with 2GB of memory and 2 threads. Loading two datasets into H2O-3 dataframes and printing their dimensions**

In [220]:
import h2o
h2o.init(nthreads = 2, max_mem_size = '2g')
trainFrameStudent = h2o.import_file(path="/Users/aishwaryaadiki/Downloads/student-mat.csv")
trainFrameNutrition = h2o.import_file(path="/Users/aishwaryaadiki/Downloads/Nutritional Data for Fast Foods.csv")
trainFrameStudent.describe()

Checking whether there is an H2O instance running at http://localhost:54321. connected.


--------------------------  -------------------------------------
H2O_cluster_uptime:         13 mins 11 secs
H2O_cluster_timezone:       America/Chicago
H2O_data_parsing_timezone:  UTC
H2O_cluster_version:        3.44.0.3
H2O_cluster_version_age:    1 month and 8 days
H2O_cluster_name:           H2O_from_python_aishwaryaadiki_pztdgv
H2O_cluster_total_nodes:    1
H2O_cluster_free_memory:    3.979 Gb
H2O_cluster_total_cores:    10
H2O_cluster_allowed_cores:  10
H2O_cluster_status:         locked, healthy
H2O_connection_url:         http://localhost:54321
H2O_connection_proxy:       {"http": null, "https": null}
H2O_internal_security:      False
Python_version:             3.11.5 final
--------------------------  -------------------------------------

Parse progress: |████████████████████████████████████████████████████████████████| (done) 100%
Parse progress: |████████████████████████████████████████████████████████████████| (done) 100%


Rows:395
Cols:31

school    sex    age                 address    famsize    Pstatus    Medu                Fedu                Mjob      Fjob      reason      guardian    traveltime          studytime           failures            schoolsup    famsup    paid    activities    nursery    higher    internet    romantic    famrel              freetime            goout               Dalc                Walc                health              absences           grade
-------  --------  -----  ------------------  ---------  ---------  ---------  ------------------  ------------------  --------  --------  ----------  ----------  ------------------  ------------------  ------------------  -----------  --------  ------  ------------  ---------  --------  ----------  ----------  ------------------  ------------------  ------------------  ------------------  ------------------  ------------------  -----------------  -------
type     enum      enum   int                 enum       enum       enum       int                 int                 enum      enum      enum        enum        int                 int                 int                 enum         enum      enum    enum          enum       enum      enum        enum        int                 int                 int                 int                 int                 int                 int                enum
mins                      15.0                                                 0.0                 0.0                                                             1.0                 1.0                 0.0                                                                                                           1.0                 1.0                 1.0                 1.0                 1.0                 1.0                 0.0
mean                      16.69620253164557                                    2.749367088607595   2.5215189873417736                                              1.4481012658227845  2.0354430379746833  0.3341772151898734                                                                                            3.9443037974683564  3.235443037974683   3.1088607594936697  1.4810126582278482  2.291139240506329   3.5544303797468384  5.708860759493671
maxs                      22.0                                                 4.0                 4.0                                                             4.0                 4.0                 3.0                                                                                                           5.0                 5.0                 5.0                 5.0                 5.0                 5.0                 75.0
sigma                     1.2760427246056265                                   1.0947351414285371  1.0882005458269433                                              0.6975047549086822  0.8392403464185556  0.7436509736062509                                                                                            0.8966586076885046  0.9988620396657203  1.1132781740183422  0.8907414280909669  1.2878965924510932  1.3903033913095773  8.003095687108184
zeros                     0                                                    3                   2                                                               0                   0                   312                                                                                                           0                   0                   0                   0                   0                   0                   115
missing  0         0      0                   0          0          0          0                   0                   0         0         0           0           0                   0                   0                   0            0         0       0             0          0         0           0           0                   0                   0                   0                   0                   0      

In [221]:
trainFrameNutrition.describe()

Rows:126
Cols:13

C1                 Fast Food Restaurant    Item                                Type                      Serving Size (g)    Calories            Total Fat (g)       Saturated Fat (g)    Trans Fat (g)       Sodium (mg)        Carbs (g)           Sugars (g)          Protein (g)
-------  -----------------  ----------------------  ----------------------------------  ------------------------  ------------------  ------------------  ------------------  -------------------  ------------------  -----------------  ------------------  ------------------  ------------------
type     int                enum                    enum                                enum                      int                 int                 real                real                 real                int                real                real                real
mins     0.0                                                                                                      44.0                130.0               3.5                 1.0                  0.0                 50.0               6.0                 0.0                 2.0
mean     62.5                                                                                                     224.26984126984127  532.4920634920635   28.177049180327874  10.153968253968255   0.821052631578947   975.6942148760328  44.572222222222216  13.280952380952384  24.92213114754097
maxs     125.0                                                                                                    467.0               1240.0              79.0                35.0                 4.0                 2460.0             106.0               93.0                69.0
sigma    36.51711927302043                                                                                        108.03306254649362  250.84429420759784  17.734058313836485  7.9600266410976355   0.9628013776374773  530.1002081835599  20.5227352519644    21.02221573207055   15.453719295003832
zeros    1                                                                                                        0                   0                   0                   0                    47                  0                  0                   17                  0
missing  0                  0                       0                                   0                         0                   0                   4                   0                    12                  5                  0                   0                   4
0        0.0                McDonald's              Hamburger                           Burger                    98.0                240.0               8.0                 3.0                  0.0                 480.0              32.0                6.0                 12.0
1        1.0                McDonald's              Cheeseburger                        Burger                    113.0               290.0               11.0                5.0                  0.5                 680.0              33.0                7.0                 15.0
2        2.0                McDonald's              ?                                   Burger                    211.0               530.0               27.0                10.0                 1.0                 960.0              47.0                9.0                 24.0
3        3.0                McDonald's              Quarter Pounder with Cheese         Burger                    202.0               520.0               26.0                12.0                 1.5                 1100.0             41.0                10.0                30.0
4        4.0                McDonald's              Bacon Clubhouse Burger              Burger                    270.0               720.0               40.0                15.0                 1.5                 1470.0             51.0                14.0                39.0
5        5.0                McDonald's              Double Qua